app/gui/view/graph-editor/src/builtin/visualization/native/text_visualization.rs

```diff
@@ -0,0 +1,469 @@
+        self.clipping_div.set_dom_size(size - 2.0 * text_padding);
+        self.size.set(size);
+    }
+
+    fn set_text_provider(&self, text_provider: T) {
+        self.text_provider.replace(Some(text_provider));
+    }
+}
+
+impl<T: TextProvider> Model<T> {
+    fn get_string_for_cell(&self, row: usize, column: usize) -> String {
+        self.text_provider
+            .borrow()
+            .as_ref()
+            .and_then(|text_provider| text_provider.get_slice(row, column))
+            .unwrap_or_default()
+    }
+}
+
+
+// ================
+// === TextGrid ===
+// ================
+
+/// Sample visualization that renders the given data as text. Useful for debugging and testing.
+#[derive(Debug, Shrinkwrap)]
+#[allow(missing_docs)]
+pub struct TextGrid<T> {
+    #[shrinkwrap(main_field)]
+    model:   Rc<Model<T>>,
+    pub frp: visualization::instance::Frp,
+    network: frp::Network,
+}
+
+impl<T: TextProvider + 'static> TextGrid<T> {
+    /// Constructor.
+    pub fn new(app: Application) -> Self {
+        let network = frp::Network::new("visualization_text_grid");
+        let frp = visualization::instance::Frp::new(&network);
+        let model = Rc::new(Model::new(app));
+        Self { model, frp, network }
+    }
+
+    fn init_frp(&self, text_provider: &dyn TextProvider) {
+        let network = &self.network;
+        let frp = &self.frp;
+        let text_provider = text_provider.frp();
+        let text_grid = &self.text_grid;
+        let model = &self.model;
+        let scrollbar_h = &self.model.scroll_bar_horizontal;
+        let scrollbar_v = &self.model.scroll_bar_vertical;
+        let dom_entry_root = &self.model.dom_entry_root;
+
+        let scene = &self.model.app.display.default_scene;
+        let style_watch = StyleWatchFrp::new(&scene.style_sheet);
+        use theme::graph_editor::visualization::text_grid as text_grid_style;
+        let font_name = style_watch.get_text(text_grid_style::font);
+        let font_size = style_watch.get_number(text_grid_style::font_size);
+
+        let longest_observed_line: Rc<Cell<u32>> = default();
+        let max_observed_lines: Rc<Cell<u32>> = default();
+
+        frp::extend! { network
+            init <- source::<()>();
+
+            on_data_update <- frp.send_data.constant(());
+            text_grid.request_model_for_visible_entries <+ any(on_data_update,init);
+
+            // === Visualisation API Inputs ===
+
+            eval frp.set_size  ((size) model.set_size(*size));
+
+
+            // === Text Grid API ===
+
+            requested_entry <- text_grid.model_for_entry_needed.map2(&text_grid.grid_size,
+                f!([model]((row, col), _grid_size) {
+                    let text = model.get_string_for_cell(*row,*col);
+                    let model = grid_view_entry::Model{text};
+                    (*row, *col, model)
+                })
+            );
+            text_grid.model_for_entry <+ requested_entry;
```

Comment:
This is really long FRP definition. Please refactor it to a few, well named and well-documented functions. See `lib/rust/ensogl/component/text/src/component/text.rs` (function `Text::init` for reference)